README.md
# Squirrel Navigator

**Squirrel Navigator** - A secure and private text-based web browser

## How to Install

You must downloading repositore of Squirrel Navigator:

> git clone <https://github.com/squirrelcom/squirrelnavigator.git>

Then you must type the command:

> python3 setup.py install

## Using

If you type **URL site** Squirrel Navigator will show site whose URL you entered.
If you type **quit** or **q** Squirrel Navigator will close.
If you type **google** or **g** Squirrel Navigator will send your request Google search engine.
If you type **wikipedia** or **w** Squirrel Navigator will send your request Wikipedia.
If you type **save_html** or **sh** to save site in format html.
If you type **save_markdown** or **smd** to save site in format markdown.


CHANGELOG.md
Changelog
=========

Version 1.0.0
-------------

- Release 1.0.0 Squirrel-Navigator has released. goodbye SquirrelWeb!! you will be missed


tests/test_akasia.py
import unittest
import sys
sys.path.insert(0, '../')
import squirrelnavigator


class Test(unittest.TestCase):

    maxDiff = None

    def test_get_request_status(self):
        request_get = squirrelnavigator.get_request('https://httpbin.org/status/200')
        self.assertEqual(request_get[1].status_code, 200)

    def test_get_request_html(self):
        request_get = squirrelnavigator.get_request('https://httpbin.org/html')
        self.assertEqual(request_get[0], '''<!DOCTYPE html>
<html>
  <head>
  </head>
  <body>
      <h1>Herman Melville - Moby-Dick</h1>

      <div>
        <p>
          Availing himself of the mild, summer-cool weather that now reigned in these latitudes, and in preparation for the peculiarly active pursuits shortly to be anticipated, Perth, the begrimed, blistered old blacksmith, had not removed his portable forge to the hold again, after concluding his contributory work for Ahab's leg, but still retained it on deck, fast lashed to ringbolts by the foremast; being now almost incessantly invoked by the headsmen, and harpooneers, and bowsmen to do some little job for them; altering, or repairing, or new shaping their various weapons and boat furniture. Often he would be surrounded by an eager circle, all waiting to be served; holding boat-spades, pike-heads, harpoons, and lances, and jealously watching his every sooty movement, as he toiled. Nevertheless, this old man's was a patient hammer wielded by a patient arm. No murmur, no impatience, no petulance did come from him. Silent, slow, and solemn; bowing over still further his chronically broken back, he toiled away, as if toil were life itself, and the heavy beating of his hammer the heavy beating of his heart. And so it was.—Most miserable! A peculiar walk in this old man, a certain slight but painful appearing yawing in his gait, had at an early period of the voyage excited the curiosity of the mariners. And to the importunity of their persisted questionings he had finally given in; and so it came to pass that every one now knew the shameful story of his wretched fate. Belated, and not innocently, one bitter winter's midnight, on the road running between two country towns, the blacksmith half-stupidly felt the deadly numbness stealing over him, and sought refuge in a leaning, dilapidated barn. The issue was, the loss of the extremities of both feet. Out of this revelation, part by part, at last came out the four acts of the gladness, and the one long, and as yet uncatastrophied fifth act of the grief of his life's drama. He was an old man, who, at the age of nearly sixty, had postponedly encountered that thing in sorrow's technicals called ruin. He had been an artisan of famed excellence, and with plenty to do; owned a house and garden; embraced a youthful, daughter-like, loving wife, and three blithe, ruddy children; every Sunday went to a cheerful-looking church, planted in a grove. But one night, under cover of darkness, and further concealed in a most cunning disguisement, a desperate burglar slid into his happy home, and robbed them all of everything. And darker yet to tell, the blacksmith himself did ignorantly conduct this burglar into his family's heart. It was the Bottle Conjuror! Upon the opening of that fatal cork, forth flew the fiend, and shrivelled up his home. Now, for prudent, most wise, and economic reasons, the blacksmith's shop was in the basement of his dwelling, but with a separate entrance to it; so that always had the young and loving healthy wife listened with no unhappy nervousness, but with vigorous pleasure, to the stout ringing of her young-armed old husband's hammer; whose reverberations, muffled by passing through the floors and walls, came up to her, not unsweetly, in her nursery; and so, to stout Labor's iron lullaby, the blacksmith's infants were rocked to slumber. Oh, woe on woe! Oh, Death, why canst thou not sometimes be timely? Hadst thou taken this old blacksmith to thyself ere his full ruin came upon him, then had the young widow had a delicious grief, and her orphans a truly venerable, legendary sire to dream of in their after years; and all of them a care-killing competency.
        </p>
      </div>
  </body>
</html>''')


if __name__ == "__main__":
    print('Version: ' + squirrelnavigator.VERSION + '\n')
    unittest.main()


akasia.py
# coding: utf-8

""" This is main module web browser Akasia. """

import sys
import requests
import html2text
import wikipedia
import dock

VERSION = '1.6.3'

# pylint settings:
# pylint: disable=E1101
# pylint: disable=E1102


@dock()
def get_request(url: str) -> str:
    """

    This function receives a request from the site.

    Args:
        url (str): A variable that stores the URL that will open in the browser Akasia.

    Returns:
        site_content (str): The variable contains the content of the site in html format.
        response (str): This variable stores the request from the site.

    """

    try:
        response = requests.get(url)
    except requests.exceptions.MissingSchema:
        choosing_the_right_url = input(
            f"Invalid URL '{url}': No schema supplied. Perhaps you meant http://{url}? (y/n) ")
        if choosing_the_right_url.lower() == 'y' or choosing_the_right_url.lower() == 'yes':
            response = requests.get(f'http://{url}')
        else:
            sys.exit()

    site_content = str(response.content, response.encoding)
    return site_content, response


@dock()
def print_site(site_content: str, response: str) -> str:
    """

    This function prints the site in format markdown.

    Args:
        site_content (str): The variable contains the content of the site in html format.
        response (str): This variable stores the request from the site.
    Returns:
        site (str): The variable stores the text of the site in markdown format.
    """
    if len(site_content) == 0:

        if response.status_code == requests.codes.ok:
            site = (html2text.html2text(site_content))
        if response.status_code == 404:
            site = ('Error 404, Not Found!')
        if response.status_code == 500:
            site = ('Error 500, Internal server error!')

        site = (html2text.html2text(site_content))

    # If non-empty content is detected, print it.
    # This is to allow customised html error messages.

    site = (html2text.html2text(site_content))
    return site


@dock()
def save_site_in_html(site_content: str, path: str) -> None:
    """

    This function is needed to save the site in html format.

    Args:
        site_content (str): This variable stores the site in html format.
        path (str): This variable stores path which will saved site in format html.

    Returns:
        None: The function returns nothing.

    """

    file = open(path, 'w')
    file.write(site_content)
    file.close()


@dock()
def save_site_in_markdown(site_content: str, path: str) -> None:
    """

    This function is needed to save the site in markdown format.

    Args:
        site_content (str): This variable stores the site in html format.
        path (str): This variable stores path which will saved site in format markdown.

    Returns:
        None: The function returns nothing.

    """
    file = open(path, 'w')
    file.write(html2text.html2text(site_content))
    file.close()


@dock()
def main() -> None:
    """ This is main function, what initializing web browser Akasia. """

    print('''
          d8888 888                        d8b          
         d88888 888                        Y8P          
        d88P888 888                                     
       d88P 888 888  888  8888b.  .d8888b  888  8888b.  
      d88P  888 888 .88P     "88b 88K      888     "88b 
     d88P   888 888888K  .d888888 "Y8888b. 888 .d888888 
    d8888888888 888 "88b 888  888      X88 888 888  888 
   d88P     888 888  888 "Y888888  88888P' 888 "Y888888\n\n\n''')
    print(f'Version - {VERSION}\n'.center(58))
    print('Akasia - A fork tiny python text-based web browser Asiakas.\n'.center(58))
    print('Type "quit" or "q" to shut down the browser.'.center(58))
    print('Type "google" or "g" to search information in Google.'.center(58))
    print('Type "wikipedia" or "w" to search information in Wikipedia.'.center(58))
    print('Type "save_html" or "sh" to save site in format html.'.center(58))
    print('Type "save_markdown" or "smd" to save site in format markdown.'.center(58))

    while True:
        link = input('URL: ')
        if link.lower() == 'quit' or link.lower() == 'q':
            break
        if link.lower() == 'google' or link.lower() == 'g':
            request = input('Request: ')
            link = ('https://google.com/search?q=' + request.replace(' ', '+'))
            cont, req_get = get_request(link)
            print(print_site(cont, req_get))
        elif link.lower() == 'wikipedia' or link.lower() == 'w':
            try:
                request = input('Request: ')
                language = input('Language on search in Wikipedia: ')
                wikipedia.set_lang(language)
                wiki_page = wikipedia.page(request)
                type_text = input('Full text(y/n) ')
                if type_text.lower() == 'y':
                    print(wiki_page.content)
                elif type_text.lower() == 'n':
                    print(wikipedia.summary(request))
                print('\nPage URL: ' + wiki_page.url)
            except wikipedia.exceptions.PageError:
                print('Request page not found')
            except requests.exceptions.ConnectionError:
                print('Please type language by first two letters in language name.')
        elif link.lower() == 'save_html' or link.lower() == 'sh':
            link = input('URL: ')
            path = input('Path: ')
            cont, req_get = get_request(link)
            save_site_in_html(cont, path)
        elif link.lower() == 'save_markdown' or link.lower() == 'smd':
            link = input('URL: ')
            path = input('Path: ')
            cont, req_get = get_request(link)
            save_site_in_markdown(cont, path)
        else:
            cont, req_get = get_request(link)
            print(print_site(cont, req_get))


if __name__ == "__main__":
    main()


asiakas.py
import requests
import html2text
import pyfiglet


def get_request(url):
    req_get = requests.get(url)
    cont = str(req_get.content, "utf-8")
    if len(cont) == 0:
        if req_get.status_code == 200:
            print(html2text.html2text(cont))
        elif req_get.status_code == 404:
            print("Error 404, Not Found!")
        elif req_get.status_code == 500:
            print("Error 500, Internal server error!")
        else:
            print(html2text.html2text(cont))

    # If non-empty content is detected, print it.
    # This is to allow customised html error messages.
    else:
        print(html2text.html2text(cont))


def main():
    fig = pyfiglet.Figlet(font='larry3d')
    print(fig.renderText('asiakas'))
    print("Asiakas - A tiny python text-based web browser\n".center(58))
    print("Type 'quit' to shut down the browser.".center(58))

    while True:
        link = input("URL: ")
        if link == "quit":
            break
        else:
            pass

        get_request(link)


if __name__ == "__main__":
    main()


squirrelnavigator.py
# coding: utf-8

""" This is main module web browser Squirrel-Navigator. """

import sys
import requests
import html2text
import wikipedia
import dock
from rich.console import Console
from rich.markdown import Markdown

VERSION = '0.1'
console = Console()

# pylint settings:
# pylint: disable=E1101
# pylint: disable=E1102


@dock()
def get_request(url: str) -> str:
    """

    This function receives a request from the site.

    Args:
        url (str): A variable that stores the URL that will open in the browser Squirrel-Navigator.

    Returns:
        site_content (str): The variable contains the content of the site in html format.
        response (str): This variable stores the request from the site.

    """

    try:
        response = requests.get(url)
    except requests.exceptions.MissingSchema:
        choosing_the_right_url = input(
            f"Invalid URL '{url}': No schema supplied. Perhaps you meant http://{url}? (y/n) ")
        if choosing_the_right_url.lower() == 'y' or choosing_the_right_url.lower() == 'yes':
            response = requests.get(f'http://{url}')
        else:
            sys.exit()

    site_content = str(response.content, response.encoding)
    return site_content, response


@dock()
def print_site(site_content: str, response: str) -> str:
    """

    This function prints the site in format markdown.

    Args:
        site_content (str): The variable contains the content of the site in html format.
        response (str): This variable stores the request from the site.
    Returns:
        site (str): The variable stores the text of the site in markdown format.
    """
    if len(site_content) == 0:

        if response.status_code == requests.codes.ok:
            site = (html2text.html2text(site_content))
        if response.status_code == 404:
            site = ('Error 404, Not Found!')
        if response.status_code == 500:
            site = ('Error 500, Internal server error!')

        site = (html2text.html2text(site_content))

    # If non-empty content is detected, print it.
    # This is to allow customised html error messages.

    site = (html2text.html2text(site_content))
    return site


@dock()
def save_site_in_html(site_content: str, path: str) -> None:
    """

    This function is needed to save the site in html format.

    Args:
        site_content (str): This variable stores the site in html format.
        path (str): This variable stores path which will saved site in format html.

    Returns:
        None: The function returns nothing.

    """

    file = open(path, 'w')
    file.write(site_content)
    file.close()


@dock()
def save_site_in_markdown(site_content: str, path: str) -> None:
    """

    This function is needed to save the site in markdown format.

    Args:
        site_content (str): This variable stores the site in html format.
        path (str): This variable stores path which will saved site in format markdown.

    Returns:
        None: The function returns nothing.

    """
    file = open(path, 'w')
    file.write(html2text.html2text(site_content))
    file.close()


@dock()
def main() -> None:
    """ This is main function, what initializing web browser Squirrel-Navigator. """

    print('''
          d8888 888                        d8b          
         d88888 888                        Y8P          
        d88P888 888                                     
       d88P 888 888  888  8888b.  .d8888b  888  8888b.  
      d88P  888 888 .88P     "88b 88K      888     "88b 
     d88P   888 888888K  .d888888 "Y8888b. 888 .d888888 
    d8888888888 888 "88b 888  888      X88 888 888  888 
   d88P     888 888  888 "Y888888  88888P' 888 "Y888888\n\n\n''')
    print(f'Version - {VERSION}\n'.center(58))
    print('Squirrel-Navigator - A private and secure text-based web browser.\n'.center(58))
    print('Type "quit" or "q" to shut down the browser.'.center(58))
    print('Type "google" or "g" to search information in Google.'.center(58))
    print('Type "wikipedia" or "w" to search information in Wikipedia.'.center(58))
    print('Type "save_html" or "sh" to save site in format html.'.center(58))
    print('Type "save_markdown" or "smd" to save site in format markdown.'.center(58))

    while True:
        link = input('URL: ')
        if link.lower() == 'quit' or link.lower() == 'q':
            break
        if link.lower() == 'google' or link.lower() == 'g':
            request = input('Request: ')
            link = ('https://google.com/search?q=' + request.replace(' ', '+'))
            cont, req_get = get_request(link)
            markdown_site = Markdown(print_site(cont, req_get))
            console.print(markdown_site)
        elif link.lower() == 'wikipedia' or link.lower() == 'w':
            try:
                request = input('Request: ')
                language = input('Language on search in Wikipedia: ')
                wikipedia.set_lang(language)
                wiki_page = wikipedia.page(request)
                type_text = input('Full text(y/n) ')
                if type_text.lower() == 'y':
                    print(wiki_page.content)
                elif type_text.lower() == 'n':
                    print(wikipedia.summary(request))
                print('\nPage URL: ' + wiki_page.url)
            except wikipedia.exceptions.PageError:
                print('Request page not found')
            except requests.exceptions.ConnectionError:
                print('Please type language by first two letters in language name.')
        elif link.lower() == 'save_html' or link.lower() == 'sh':
            link = input('URL: ')
            path = input('Path: ')
            cont, req_get = get_request(link)
            save_site_in_html(cont, path)
        elif link.lower() == 'save_markdown' or link.lower() == 'smd':
            link = input('URL: ')
            path = input('Path: ')
            cont, req_get = get_request(link)
            save_site_in_markdown(cont, path)
        else:
            cont, req_get = get_request(link)
            markdown_site = Markdown(print_site(cont, req_get))
            console.print(markdown_site)


if __name__ == "__main__":
    main()
